Sprint 6 solution

PavelAntipov opened the pull request:
изменён history manager согласно ТЗ 6-го спринта
доработан вывод toString для эпиков, тасок и сабтасок
пофикшены мелкие баги, которые вылезали при удалении сабтасок (добавлено удаление на по индексу, а по значению элемента)
написаны тесты для history manager

PavelAntipov committed:
1. Class Node added
2. methods getHistory and add was modified in class InMemoryHistoryManager
3. method remove was modified in class InMemoryHistoryManager

PavelAntipov committed:
1. InMemoryHistoryManager bugs fixed
2. Optional scenario for customer in Main was added
3. Tests for InMemoryHistoryManager was added (task,subtask,epic deletion and history accounting)
4. toString for epic, task and subtask was modified

PavelAntipov committed:
code analysis warnings was reviewed, some unused code and import statements was deleted

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

PavelAntipov committed:
fix test epicCantBeAddedAsSubtaskToItself

PavelAntipov committed:
delete non-actual history manager tests

PavelAntipov committed:
Fixes for TaskManagerTest tests

AlexeyCheuzov commented on src/issues/EpicTest.java:
```diff
@@ -5,22 +5,22 @@
 import static org.junit.jupiter.api.Assertions.*;
 
 class EpicTest {
```
Все тесты должны быть в папке с тестами

PavelAntipov replied:
fixed

AlexeyCheuzov commented on src/InMemoryHistoryManager.java:
```diff
@@ -1,24 +1,124 @@
-    private final LinkedList<Task> history = new LinkedList<>();
-    private static final int MAX_SIZE = 10;
+public class InMemoryHistoryManager implements HistoryManager {
+    Map<Integer, Node<Task>> listOfTasks = new HashMap<>();
```
Map реализована не совсем верно. Должно быть две ноды, head и last, а map нужна для того, чтобы производить операции по удалению элементов за O(1). Для примера можно посмотреть LinkedList

AlexeyCheuzov commented on src/InMemoryHistoryManager.java:
```diff
@@ -1,24 +1,124 @@
-    private final LinkedList<Task> history = new LinkedList<>();
-    private static final int MAX_SIZE = 10;
+public class InMemoryHistoryManager implements HistoryManager {
+    Map<Integer, Node<Task>> listOfTasks = new HashMap<>();
```
У всех переменных должны быть модификаторы доступа

PavelAntipov replied:
fixed

AlexeyCheuzov commented on src/InMemoryHistoryManager.java:
```diff
@@ -1,24 +1,124 @@
     @Override
-    public List<Task> getHistory(){
-        return new ArrayList<>(history);
+    public void add(Task task) {
```
Метод add соответственно поменяется. Сначала должна удаляться повторяющаяся нода, если она есть(можно также взять пример с linkedList. у двух соседних нод меняются ссылки). Затем новую нода привязывается как last нода и добавляется в map. При этом необходимо проверить не является ли новая задача null'ом. Желательно разделить эту логику на несколько методов(remove, linkLast)

PavelAntipov replied:
готово

AlexeyCheuzov commented on src/InMemoryTaskManager.java:
```diff
@@ -14,7 +14,7 @@ public class InMemoryTaskManager implements TaskManager {
     protected Map<Integer, Epic> epics = new HashMap<>();
     //Метод для получения ID новой задачи
-    private HistoryManager historyManager;
+    private final HistoryManager historyManager;
```
При удалении задач, они также должны удаляться из historyManager (методы deleteSubtasks, deleteEpics и deleteTasks в том числе)

PavelAntipov replied:
поправил

PavelAntipov committed:
fixes after 1st code review:
1)history manager logic totally changed
2)ammended methods deleteEpics(), deleteSubtasks(), deleteTasks with historyManager functions
3)moved all test cases in Test folder

PavelAntipov committed:
code style fixes

PavelAntipov committed:
code style fixes

AlexeyCheuzov commented on src/Node.java:
```diff
@@ -0,0 +1,50 @@
+
+    public Node(T data) {
+        this.data = data;
+        this.head = null;
```
инициализировать значения через null нет особого смысла, так как это значение по дефолту

PavelAntipov replied:
поправил

AlexeyCheuzov commented on src/InMemoryHistoryManager.java:
```diff
@@ -1,24 +1,99 @@
-       history.add(task);
+    private boolean taskIsNull(Task task) {
+        boolean result = false;
+        if (task.equals(null)) {
```
Проверка не совсем корректная, так как если объект null то у него не будет никаких методов, и при попытке его вызывать произойдет NPE

PavelAntipov replied:
хотелось покрасивее. поправил

AlexeyCheuzov commented on src/InMemoryHistoryManager.java:
```diff
@@ -1,24 +1,99 @@
+        }
+        Node<Task> newNode = new Node<>(task);
+        if (listOfTasks.isEmpty()) { //мапа пустая
+            listOfTasks.put(task.getId(), newNode);
```
В методе есть дублирующийся код, можно вынести его либо в отдельный метод, либо в addLast. Советую посмотреть как работает метод linkLast у LinkedList, можно опираться на него. Также советую в будущем стараться избегать большого количества if else там где это возможно, особенно если они вложенные, так как это может затруднить чтение кода.

PavelAntipov replied:
вот это спасибо. поменял код, очень красиво получилось

PavelAntipov committed:
1)addLast in InMemoryHistoryManager was modified
2)new constructor for Node was added
